compiler/z/codegen/OMRMachine.cpp

```diff
@@ -5327,40 +5327,70 @@ OMR::Z::Machine::coerceRegisterAssignment(TR::Instruction
-
-         spareReg->setState(TR::RealRegister::Assigned);
-         spareReg->setAssignedRegister(currentTargetVirtual);
-         currentTargetVirtual->setAssignedRegister(spareReg);
+         // if spareReg is still NULL, then it means we did not find any free Reg, and no
+         // other register is available to spill. So we must spill the target register.
+
+         // In general if a register is blocked, the virtual register
+         // associated with a target register would have been assigned as part of that instruction.
+         // So we would not do this (i.e. spill the blocked register). However, in this case all
+         // pre-assigned registers in the dependency are blocked before individiual dependencies are coerced
+         // into their respective target registers (inside TR_S390RegisterDependencyGroup::assignRegisters((..)).
+         // In this situation it's possible that the target register of a register dependency is currently
+         // blocked (as in here) and occupied by another virtual register and that there is no other available
+         // register to free.
+         // Hence, we must spill the blocked target reg as it is the only available candidate.
+         if (spareReg == NULL)
+            {
+            self()->cg()->setRegisterAssignmentFlag(TR_RegisterSpilled);
+            virtualRegister->block();
 
-         if (enableHighWordRA && currentTargetVirtual->is64BitReg())
-            {
-            //TR_ASSERT(targetRegisterHW->getState() == TR::RealRegister::Blocked,
-            //        "currentTargetVirtual is blocked and is fullsize, but the HW is not blocked?");
+            self()->spillRegister(currentInstruction, currentTargetVirtual);
+            targetRegister->setAssignedRegister(virtualRegister);
+            virtualRegister->setAssignedRegister(targetRegister);
+            targetRegister->setState(TR::RealRegister::Assigned);
 
-            spareReg->getHighWordRegister()->setState(TR::RealRegister::Assigned);
-            spareReg->getHighWordRegister()->setAssignedRegister(currentTargetVirtual);
-            targetRegister->getLowWordRegister()->setState(TR::RealRegister::Unlatched);
-            targetRegister->getLowWordRegister()->setAssignedRegister(NULL);
-            targetRegisterHW->setState(TR::RealRegister::Unlatched);
-            targetRegisterHW->setAssignedRegister(NULL);
-            }
-         if (virtualRegister->getTotalUseCount() != virtualRegister->getFutureUseCount())
-            {
-            self()->cg()->setRegisterAssignmentFlag(TR_RegisterReloaded);
-            self()->reverseSpillState(currentInstruction, virtualRegister, targetRegister);
+            if (targetRegister->isHighWordRegister() && currentTargetVirtual->is64BitReg() &&
+                targetRegister->getRegisterNumber() == spareReg->getHighWordRegister()->getRegisterNumber())
+               doNotRegCopy = true;
```

Comment:
@dchopra001 this looks incorrect. `spareReg` is `NULL` so we have a null pointer dereference here. What is the intention of this `if` block? I do not see any path following this block which leads to a use of `doNotRegCopy`.